File: apps/desktop-tauri/src-tauri/src/ssh_config.rs
//! `list_ssh_config_hosts`: the concrete `Host` entries of `~/.ssh/config`
//! (with one level of `Include`), so the Remote SSH page can offer them as
//! targets. Only the fields the picker shows are extracted; `ssh` itself
//! resolves everything else when the alias is used.

use std::path::{Path, PathBuf};

use serde::Serialize;
use serde_json::Value;
use tauri::{AppHandle, Manager, Runtime};

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SshConfigHost {
    pub alias: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identity_file: Option<String>,
}

pub fn list_ssh_config_hosts<R: Runtime>(app: &AppHandle<R>) -> Result<Value, String> {
    let Ok(home) = app.path().home_dir() else {
        return Ok(Value::Array(Vec::new()));
    };
    let ssh_dir = home.join(".ssh");
    let hosts = parse_config_file(&ssh_dir.join("config"), &ssh_dir, &home, true);
    serde_json::to_value(hosts).map_err(|e| e.to_string())
}

/// Parses one config file. `follow_includes` is cleared for included files
/// (one level of `Include`, as the picker promises).
pub fn parse_config_file(
    path: &Path,
    ssh_dir: &Path,
    home: &Path,
    follow_includes: bool,
) -> Vec<SshConfigHost> {
    let Ok(contents) = std::fs::read_to_string(path) else {
        return Vec::new();
    };
    parse_config(&contents, ssh_dir, home, follow_includes)
}

fn parse_config(
    contents: &str,
    ssh_dir: &Path,
    home: &Path,
    follow_includes: bool,
) -> Vec<SshConfigHost> {
    let mut hosts: Vec<SshConfigHost> = Vec::new();
    // `None` while outside any block or inside a `Match` block.
    let mut current: Option<Vec<SshConfigHost>> = None;

    for raw_line in contents.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let Some((keyword, value)) = split_keyword(line) else {
            continue;
        };
        match keyword.to_ascii_lowercase().as_str() {
            "host" => {
                flush(&mut hosts, current.take());
                let aliases: Vec<SshConfigHost> = split_values(value)
                    .into_iter()
                    .filter(|alias| !alias.contains(['*', '?', '!']))
                    .map(|alias| SshConfigHost {
                        alias,
                        ..Default::default()
                    })
                    .collect();
                current = Some(aliases);
            }
            "match" => {
                flush(&mut hosts, current.take());
            }
            "include" => {
                if follow_includes {
                    for pattern in split_values(value) {
                        for included in resolve_include(&pattern, ssh_dir, home) {
                            let entries = parse_config_file(&included, ssh_dir, home, false);
                            match current.as_mut() {
                                // Inside a Host block an Include contributes options to that block.
                                Some(block) => {
                                    for entry in entries {
                                        for host in block.iter_mut() {
                                            merge_missing(host, &entry);
                                        }
                                    }
                                }
                                None => hosts.extend(entries),
                            }
                        }
                    }
                }
            }
            other => {
                if let Some(block) = current.as_mut() {
                    let value = unquote(value);
                    for host in block.iter_mut() {
                        apply_option(host, other, &value);
                    }
                }
            }
        }
    }
    flush(&mut hosts, current.take());
    dedupe(hosts)
}

fn flush(hosts: &mut Vec<SshConfigHost>, block: Option<Vec<SshConfigHost>>) {
    if let Some(block) = block {
        hosts.extend(block);
    }
}

fn apply_option(host: &mut SshConfigHost, keyword: &str, value: &str) {
    if value.is_empty() {
        return;
    }
    // ssh keeps the first value it sees for an option.
    match keyword {
        "hostname" if host.host_name.is_none() => host.host_name = Some(value.to_string()),
        "user" if host.user.is_none() => host.user = Some(value.to_string()),
        "port" if host.port.is_none() => host.port = value.parse::<u16>().ok().filter(|p| *p > 0),
        "identityfile" if host.identity_file.is_none() => {
            host.identity_file = Some(value.to_string())
        }
        _ => {}
    }
}

fn merge_missing(host: &mut SshConfigHost, other: &SshConfigHost) {
    if host.host_name.is_none() {
        host.host_name.clone_from(&other.host_name);
    }
    if host.user.is_none() {
        host.user.clone_from(&other.user);
    }
    if host.port.is_none() {
        host.port = other.port;
    }
    if host.identity_file.is_none() {
        host.identity_file.clone_from(&other.identity_file);
    }
}

/// First occurrence of an alias wins, matching ssh's first-match semantics.
fn dedupe(hosts: Vec<SshConfigHost>) -> Vec<SshConfigHost> {
    let mut seen = std::collections::HashSet::new();
    hosts
        .into_iter()
        .filter(|host| seen.insert(host.alias.clone()))
        .collect()
}

/// `Keyword value`, `Keyword=value`, or `Keyword = value`.
fn split_keyword(line: &str) -> Option<(&str, &str)> {
    let end = line.find(|c: char| c.is_whitespace() || c == '=')?;
    let keyword = &line[..end];
    let rest = line[end..]
        .trim_start_matches(|c: char| c.is_whitespace() || c == '=')
        .trim();
    (!keyword.is_empty()).then_some((keyword, rest))
}

fn unquote(value: &str) -> String {
    let trimmed = value.trim();
    if trimmed.len() >= 2 && trimmed.starts_with('"') && trimmed.ends_with('"') {
        trimmed[1..trimmed.len() - 1].to_string()
    } else {
        trimmed.to_string()
    }
}

/// Whitespace-separated values, honouring double quotes.
fn split_values(value: &str) -> Vec<String> {
    let mut values = Vec::new();
    let mut current = String::new();
    let mut quoted = false;
    for c in value.chars() {
        match c {
            '"' => quoted = !quoted,
            c if c.is_whitespace() && !quoted => {
                if !current.is_empty() {
                    values.push(std::mem::take(&mut current));
                }
            }
            c => current.push(c),
        }
    }
    if !current.is_empty() {
        values.push(current);
    }
    values
}

fn resolve_include(pattern: &str, ssh_dir: &Path, home: &Path) -> Vec<PathBuf> {
    let expanded = if let Some(rest) = pattern.strip_prefix("~/") {
        home.join(rest)
    } else if pattern == "~" {
        home.to_path_buf()
    } else if Path::new(pattern).is_absolute() {
        PathBuf::from(pattern)
    } else {
        ssh_dir.join(pattern)
    };
    let text = expanded.to_string_lossy();
    match glob::glob(&text) {
        Ok(paths) => {
            let mut files: Vec<PathBuf> = paths
                .filter_map(Result::ok)
                .filter(|p| p.is_file())
                .collect();
            files.sort();
            files
        }
        Err(_) => Vec::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn host(alias: &str) -> SshConfigHost {
        SshConfigHost {
            alias: alias.into(),
            ..Default::default()
        }
    }

    #[test]
    fn parses_hosts_includes_aliases_and_skips_wildcards_and_match() {
        let home = tempfile::tempdir().unwrap();
        let ssh_dir = home.path().join(".ssh");
        std::fs::create_dir_all(ssh_dir.join("conf.d")).unwrap();
        std::fs::write(
            ssh_dir.join("conf.d/work.conf"),
            "Host build\n  HostName build.internal\n  User ci\n  Port 2222\n  IdentityFile ~/.ssh/id_work\n\nInclude nested.conf\n",
        )
        .unwrap();
        // Nested includes are not followed (one level only).
        std::fs::write(
            ssh_dir.join("nested.conf"),
            "Host nested\n  HostName nested.example\n",
        )
        .unwrap();
        let config = r#"
# comment
Include conf.d/*.conf

Host dev staging
  HostName=dev.example.com
  User "alice"
  Port 22

Host *.example.com !dev bastion?
  User nobody

Match host build
  User matched

Host lab
  HostName lab.local
  HostName ignored.later

Host dev
  User duplicate
"#;
        let hosts = parse_config(config, &ssh_dir, home.path(), true);
        assert_eq!(
            hosts,
            vec![
                SshConfigHost {
                    host_name: Some("build.internal".into()),
                    user: Some("ci".into()),
                    port: Some(2222),
                    identity_file: Some("~/.ssh/id_work".into()),
                    ..host("build")
                },
                SshConfigHost {
                    host_name: Some("dev.example.com".into()),
                    user: Some("alice".into()),
                    port: Some(22),
                    ..host("dev")
                },
                SshConfigHost {
                    host_name: Some("dev.example.com".into()),
                    user: Some("alice".into()),
                    port: Some(22),
                    ..host("staging")
                },
                SshConfigHost {
                    host_name: Some("lab.local".into()),
                    ..host("lab")
                },
            ]
        );
    }

    #[test]
    fn missing_config_is_empty() {
        let home = tempfile::tempdir().unwrap();
        let ssh_dir = home.path().join(".ssh");
        assert!(parse_config_file(&ssh_dir.join("config"), &ssh_dir, home.path(), true).is_empty());
    }

    #[test]
    fn splits_keywords_and_values() {
        assert_eq!(split_keyword("HostName=x"), Some(("HostName", "x")));
        assert_eq!(split_keyword("Port = 22"), Some(("Port", "22")));
        assert_eq!(split_values(r#"a "b c" d"#), vec!["a", "b c", "d"]);
    }
}

File: apps/desktop/src/integrations/ssh-config.ts
import { glob, readFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

export interface SshConfigHost {
  alias: string;
  hostName?: string;
  user?: string;
  port?: number;
  identityFile?: string;
  proxyJump?: string;
}
export interface SshConfigFiles {
  read(file: string): Promise<string | null>;
  matches(pattern: string): Promise<string[]>;
}

function words(value: string): string[] {
  return (value.match(/"[^"]*"|'[^']*'|[^\s#]+/g) ?? []).map((word) =>
    word.replace(/^(["'])(.*)\1$/, "$2"),
  );
}

interface Directive {
  keyword: string;
  values: string[];
}
function directive(raw: string): Directive | null {
  const line = raw.trim();
  if (!line || line.startsWith("#")) return null;
  const match = /^([^\s=]+)[\s=]+(.*)$/.exec(line);
  if (!match) return null;
  return { keyword: match[1]!.toLowerCase(), values: words(match[2]!) };
}
function applyProperty(host: SshConfigHost, keyword: string, value: string): void {
  if (keyword === "hostname") host.hostName ??= value;
  if (keyword === "user") host.user ??= value;
  if (keyword === "identityfile") host.identityFile ??= value;
  if (keyword === "proxyjump") host.proxyJump ??= value;
  if (keyword !== "port" || !/^\d+$/.test(value)) return;
  const port = Number(value);
  if (port > 0 && port <= 65535) host.port ??= port;
}
function mergeMetadata(targets: SshConfigHost[], entries: SshConfigHost[]): void {
  for (const entry of entries) {
    for (const host of targets) {
      host.hostName ??= entry.hostName;
      host.user ??= entry.user;
      host.port ??= entry.port;
      host.identityFile ??= entry.identityFile;
      host.proxyJump ??= entry.proxyJump;
    }
  }
}
function expandInclude(home: string, pattern: string): string {
  if (pattern.startsWith("~/")) return path.join(home, pattern.slice(2));
  if (path.isAbsolute(pattern)) return pattern;
  return path.join(home, ".ssh", pattern);
}
/** Picker metadata only: OpenSSH evaluates aliases, Match, credentials and defaults. */
export async function readSshConfigHosts(
  home: string,
  files: SshConfigFiles,
): Promise<SshConfigHost[]> {
  async function includedHosts(patterns: string[]): Promise<SshConfigHost[]> {
    const entries: SshConfigHost[] = [];
    for (const pattern of patterns) {
      const matches = await files.matches(expandInclude(home, pattern));
      for (const included of matches.sort()) entries.push(...(await parse(included, false)));
    }
    return entries;
  }
  async function parse(file: string, includes: boolean): Promise<SshConfigHost[]> {
    const text = await files.read(file);
    if (text === null) return [];
    const hosts: SshConfigHost[] = [];
    let current: SshConfigHost[] | null = null;
    for (const raw of text.split(/\r?\n/)) {
      const parsed = directive(raw);
      if (!parsed) continue;
      const { keyword, values } = parsed;
      if (keyword === "host" || keyword === "match") {
        if (current) hosts.push(...current);
        current =
          keyword === "host"
            ? values.filter((alias) => !/[*?!]/.test(alias)).map((alias) => ({ alias }))
            : null;
        continue;
      }
      if (keyword === "include" && includes) {
        const entries = await includedHosts(values);
        if (current) mergeMetadata(current, entries);
        else hosts.push(...entries);
        continue;
      }
      const value = values[0];
      if (!current || !value) continue;
      for (const host of current) applyProperty(host, keyword, value);
    }
    if (current) hosts.push(...current);
    return hosts;
  }
  const unique = new Map<string, SshConfigHost>();
  for (const host of await parse(path.join(home, ".ssh", "config"), true)) {
    if (!unique.has(host.alias)) unique.set(host.alias, host);
  }
  return [...unique.values()];
}

export async function listSshConfigHosts(): Promise<SshConfigHost[]> {
  return readSshConfigHosts(os.homedir(), {
    async read(file) {
      try {
        return await readFile(file, "utf8");
      } catch {
        return null;
      }
    },
    async matches(pattern) {
      const matches: string[] = [];
      try {
        for await (const file of glob(pattern)) matches.push(file);
      } catch {
        /* Unreadable includes add no picker entries. */
      }
      return matches;
    },
  });
}
